app_utils/hugging_face_utils.py

```diff
@@ -0,0 +1,144 @@
+    card = huggingface_hub.ModelCard.from_template(
+        card_data,
+        template_path="model_card_template.md",
+        base_model=cfg.llm_backbone,  # will be replaced in template if it exists
+        repo_id=repo_id,
+        model_architecture=model.backbone.__repr__(),
+        config=cfg.__repr__(),
+        use_fast=cfg.tokenizer.use_fast,
+        min_new_tokens=cfg.prediction.min_length_inference,
+        max_new_tokens=cfg.prediction.max_length_inference,
+        do_sample=cfg.prediction.do_sample,
+        num_beams=cfg.prediction.num_beams,
+        temperature=cfg.prediction.temperature,
+        repetition_penalty=cfg.prediction.repetition_penalty,
+        text_prompt_start=cfg.dataset.text_prompt_start,
+        text_answer_separator=cfg.dataset.text_answer_separator,
+        trust_remote_code=cfg.environment.trust_remote_code,
+        transformers_version=transformers.__version__,
+        accelerate_version=accelerate.__version__,
+        torch_version=torch.__version__.split("+")[0],
+        end_of_sentence=cfg._tokenizer_eos_token
+        if cfg.dataset.add_eos_token_to_prompt
+        else "",
+    )
+    return card
+
+####################################################################################
+#
+# Method to publish the model to Hugging Face
+#
+# experiment_path -> The path where the files of the fine-tuned model are located
+# device -> 'cpu' or 'cuda:0', if the GPU device id is 0
+# api_key -> The Hugging Face API Key
+# repo_id -> The Hugging Face Repository ID
+#
+#####################################################################################
+
+def publish_model_to_hugging_face(experiment_path: str, device: str, api_key: str, repo_id: str) -> None:
+
+  cfg, model, tokenizer = load_cfg_model_tokenizer(
+                            experiment_path,
+                            merge=True,
+                            device=device,
+                          )
+
+  check_disk_space(model.backbone, "./")
+
+  huggingface_hub.login(api_key)
+
+  user_id = huggingface_hub.whoami()["name"]
+
+  # push tokenizer to hub
+  tokenizer.push_to_hub(repo_id=repo_id, private=True)
+
+
+  # push model card to hub
+  card = get_model_card(cfg, model, repo_id)
+  card.push_to_hub(
+          repo_id=repo_id, repo_type="model", commit_message="Upload model card"
+  )
+
+  # push config to hub
+  api = huggingface_hub.HfApi()
+  api.upload_file(
+        path_or_fileobj=f"{experiment_path}/cfg.yaml",
+        path_in_repo="cfg.yaml",
+        repo_id=repo_id,
+        repo_type="model",
+        commit_message="Upload cfg.yaml",
+  )
+
+  # push model to hub
+  model.backbone.config.custom_pipelines = {
+        "text-generation": {
+            "impl": "h2oai_pipeline.H2OTextGenerationPipeline",
+            "pt": "AutoModelForCausalLM",
+        }
+  }
+
+  model.backbone.push_to_hub(
+      repo_id=repo_id,
+      private=True,
+      commit_message="Upload model",
+      safe_serialization=False,
```

Comment:
we need the safe serialization functionality here

Reply:
safe_serialization=True now